### DISCUSSION TITLE
How to add rust code in the sub-module of python project?

### DISCUSSION THREAD

**Discussion by evan0greenup** [PyO3/pyo3#2909] [Questions]
If I want to add my rust code to the sub-module (not the top level) of my python project, what should I do?

For example I the root module of my python project is `hello`. It contains three sub-module `hello.alice`, `hello.bob`, `hello.carol`. I want my rust code define the member attribute (fields and functions) of module `hello.alice`. And the other two sub-module is pure-python. What should I do?

**Comment by shenjackyuanjie** [comment]
## English is not my native so sorry if there is some mistake
# normal way
as you can see, pyo3 project have a `__init__.py` file
you can try use it to make your rust code a sub module of your python project
like this
```
- hello:
  - __init__.py
  - rust_code:
    - __init__.py
    - some.cp38-win_amd64.pyd
  - some_other.py
```
And remeber you need to copy your `pyd/so` build file in to `rust_code` dict
and remeber to write some usefull thing in your `rust_code/__init__.py`
```python3
from .some import *    # remeber to use releative import
from typing import TYPE_CHECKING
if TYPE_CHECKING:
  def this_is_rust(some_int: int) -> str: ...
```
Then you can use your rust code
```python3
from hello.rust_code import this_is_rust
this_is_rust(1123)
```
```python3
from hello import rust_code
rust_code.this_is_rust(1234)
```

# tricky way
If you dont want to copy `pyd/so` in your project dir, you can use `__init__.py` to do import for you
like this
```
- hello:
  - rust_code:
    - __init__.py
  - some_other.py
- libs
  - __init__.py
  - some.cp38-win_amd64.pyd
```
`hello/rust_code/__init__.py`
```python3
from libs import *
```
`libs/__init__.py`
```python3
from .some import *    # remeber to use releative import
```
Then you can add these typing code to any of them
```python3
from typing import TYPE_CHECKING
if TYPE_CHECKING:
  def this_is_rust(some_int: int) -> str: ...
```
Then you can use your code just like normal way
(you can see some example in [my project](https://github.com/shenjackyuanjie/Difficult-Rocket/tree/main/libs/Difficult_Rocket_rs))
